### ISSUE TITLE
Not able to install pe-parse in Windows

### ISSUE THREAD

**Issue by t3mp1337** [trailofbits/pe-parse#50]
I am trying to install this library on Windows without success.

I installed cmake version for windows from:
https://cmake.org/download/

After that I add problem of 'Unable to find vcvarsall.bat'.
I fixed it by running:
`SET VS90COMNTOOLS=%VS140COMNTOOLS%`

This according to [this](http://stackoverflow.com/questions/2817869/error-unable-to-find-vcvarsall-bat):

> Execute the following command based on the version of Visual Studio installed:
> 
> Visual Studio 2010 (VS10): SET VS90COMNTOOLS=%VS100COMNTOOLS%
> Visual Studio 2012 (VS11): SET VS90COMNTOOLS=%VS110COMNTOOLS%
> Visual Studio 2013 (VS12): SET VS90COMNTOOLS=%VS120COMNTOOLS%
> Visual Studio 2015 (VS14): SET VS90COMNTOOLS=%VS140COMNTOOLS%

Now I have other problem that it doesn't find the file: 'pepy.cpp':

> c:\>C:\Python27\python.exe C:\Python27\pe-parse-master\python\setup.py build
> running build
> running build_ext
> building 'pepy' extension
> creating build
> creating build\temp.win32-2.7
> creating build\temp.win32-2.7\Release
> creating build\temp.win32-2.7\parser-library
> creating build\temp.win32-2.7\Release\Python27
> creating build\temp.win32-2.7\Release\Python27\pe-parse-master
> creating build\temp.win32-2.7\Release\Python27\pe-parse-master\parser-library
> creating build\temp.win32-2.7\Release\Python27\pe-parse-master\parser-library\python
> C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -I/usr/local/include -I/opt/local/include -I/usr/include -I../parser-library -IC:\Python27\include -IC:\Python27\PC /Tppepy.cpp /Fobuild\temp.win32-2.7\Release\pepy.obj -g -O0
> cl : Command line warning D9002 : ignoring unknown option '-g'
> cl : Command line warning D9002 : ignoring unknown option '-O0'
> pepy.cpp
> c1xx: fatal error C1083: Cannot open source file: **'pepy.cpp': No such file or directory**
> error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\cl.exe' failed with exit status 2

I tried to edit the setp.py file like that:
```
sources = ['pepy.cpp',
                  '../parser-library/parse.cpp',
                  '../parser-library/buffer.cpp',
		   r'C:\Python27\pe-parse-master\parser-library\python\pepy.cpp',
		   r'C:\Python27\pe-parse-master\parser-library\parse.cpp',
		   r'C:\Python27\pe-parse-master\parser-library\buffer.cpp'
 ]
```

Any idea ?

**Comment by morrowd** [comment]
Are you still having issues installing / building on Windows? I was able to install / build on Windows 8.1 Pro with Visual Studio 2015 and CMAKE.

**Comment by jacwil** [comment]
I think this is specific to compiling the Python 2.7 bindings in Windows. I have this issue too but the Python 3.5 bindings in Windows compile and install cleanly without issue or ad-hoc steps.

**Comment by vamsi-una** [comment]
The following error persists:

running build
running build_ext
building 'pepy' extension
error: [WinError 2] The system cannot find the file specified

Tried with python 2.7 & python 3.5 with VS 2017.

**Comment by jacwil** [comment]
@vamsi-una can you checkout my [**win_python** branch](https://github.com/jacwil/pe-parse/tree/win_python) and see if it fixes Win 7?

**Comment by vamsi-una** [comment]
The following problem goes once you use python 3.5 or define the mentioned Env var :
`Executing task: python setup.py build ; python setup.py install <

running build
running build_ext
building 'pepy' extension
error: Unable to find vcvarsall.bat
running install
running build
running build_ext
building 'pepy' extension
error: Unable to find vcvarsall.bat
The terminal process terminated with exit code: 1`

but the path problem is still not resolved tried your code as well. 
Ps: I have VS 2017 community edition.

you updated setup.py but have you tried in source if there are other path mentioned. Definitely it seems problem of OS (win7).

**Comment by jacwil** [comment]
Interesting. I don't see any obvious paths to files inside the **python** directory outside of setup.py. I'm not sure what's going on here for Win 7.

There is the "/" in `#include <parser-library/parse.h>` of pepy.cpp, but recent versions of Visual Studio are good at honoring "/" as a directory separator. That and pepy.cpp is the file that appears to be failing to be found so it wouldn't be getting parsed yet anyways.

Using my branch, you should be given absolute paths to the source files. Hopefully this will help in troubleshooting exactly what file path is being attempted.

I renamed pepy.cpp to pepyy.cpp and ran the build from my branch:
> ...
> C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -IC:\Python27\include -IC:\src\pe-parse\pe-parser-library\include -IC:\usr\include -IC:\Python27\include -IC:\Python27\PC **/TpC:\src\pe-parse\python\pepy.cpp** /Fobuild\temp.win-amd64-2.7\Release\src\pe-parse\python\pepy.obj /EHsc
pepy.cpp
> c1xx: fatal error C1083: Cannot open source file: **'C:\src\pe-parse\python\pepy.cpp': No such file or directory**
> error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\cl.exe' failed with exit status 2

**Comment by vamsi-una** [comment]
yeah header VS can parse.
i added following in setup.py:

> print (SOURCE_FILES)
> print (INCLUDE_DIRS)
> print (LIBRARY_DIRS)
> 

and the output is :

> python setup.py build


print (SOURCE_FILES)

> ['C:\\Users\\XX\\Desktop\\try\\pe-parse-master\\pe-parse-master\\python\\pepy.cpp', 'C:\\Users\\XX\\Desktop\\try\\pe-parse-master\\pe-parse-master\\pe-parser-library\\src\\parse.cpp', 'C:\\Users\\XX\\Desktop\\try\\pe-parse-master\\pe-parse-master\\pe-parser-library\\src\\buffer.cpp']


print (INCLUDE_DIRS)

> ['C:\\Program Files\\Python 3.5\\include', 'C:\\Users\\XX\\Desktop\\try\\pe-parse-master\\pe-parse-master\\pe-parser-library\\include', 'C:\\usr\\include']


print (LIBRARY_DIRS)

> ['C:\\Program Files\\Python 3.5\\libs', 'C:\\usr\\lib']
> 
> running build
> running build_ext
> building 'pepy' extension
> error: [WinError 2] The system cannot find the file specified
> 

you see problem..!!!!!!!!

**Comment by jacwil** [comment]
I don't see anything obvious. I do find it unusual that the error doesn't print the filename.

I know that spaces in paths and non-English characters in paths (Unicode) is sometimes a problem for programs.

If your username contains non-English characters, can you try copying `C:\Users\XX\Desktop\try\pe-parse-master` to `C:\src\pe-parse-master`? This will also remove the possibility that `MAXPATH` (260 characters) doesn't get exceeded, which can also be a problem.

Likewise, uninstalling Python 3.5 and then reinstalling Python 3.5 to `C:\Python35` is something else that can be tried to get rid of the spaces found in "Program Files".

If any of the above fixes the issue for you, we can hopefully either fix this in setup.py or at the minimum check for it. The check would print a useful error message. Example: "Non-ASCII characters in file path are known to cause build errors."

**Comment by vamsi-una** [comment]
My username doesn't have any non-english or unicode character.
Max path error would have given us a different error code.

Spaces of program files doesn't come into picture for pepy.cpp......can u
give it a try on win 7..?

That way u can have a fresh look.

**Comment by jacwil** [comment]
I'll see if I can get Windows 7 installed on a machine and try. I'm sure I have an install disk laying around somewhere. Thank you for providing feedback with this issue!

**Comment by jacwil** [comment]
I updated pe-parser-library/cmake/peparse-config.cmake to support versions of Windows prior to Win10. This didn't seem to affect building and installing both Python 2.7 and Python 3.5 bindings. I was successful before and after the change.

Tested on:
Windows 7 64bit (SP1)
Visual Studio 2017 Community Edition
CMake 3.12.2

I can't repro the build errors you are seeing when using the `https://github.com/jacwil/pe-parse` fork's `win_python` branch.
